core/src/repair_weight.rs

```diff
@@ -249,28 +271,54 @@ impl RepairWeight {
     }
 
     /// Split `slot` and descendants into an orphan tree in repair weighting
-    /// These orphaned slots should be removed from `unrooted_slots` as on proper repair these slots might
-    /// now be part of the rooted path
+    /// Even if these orphaned slots were previously pruned, they should be added back properly as
+    /// on repair they might now be part of the rooted path.
+    /// Assumes that `slot` is greater than `self.root`
     pub fn split_off(&mut self, slot: Slot) {
+        assert!(slot >= self.root);
         if slot == self.root {
             error!("Trying to orphan root of repair tree {}", slot);
             return;
         }
-        if let Some(subtree_root) = self.slot_to_tree.get(&slot) {
+        if let Some(TreeRoot::Root(subtree_root)) = self.slot_to_tree.get(&slot) {
             if *subtree_root == slot {
                 info!("{} is already orphan, skipping", slot);
                 return;
             }
             let subtree = self
                 .trees
                 .get_mut(subtree_root)
-                .expect("subtree must exist");
+                .expect("`self.slot_to_tree` and `self.trees` must be in sync");
             let orphaned_tree = subtree.split_off(&(slot, Hash::default()));
-            for ((orphaned_slot, _), _) in orphaned_tree.all_slots_stake_voted_subtree() {
-                self.unrooted_slots.remove(orphaned_slot);
-                self.slot_to_tree.insert(*orphaned_slot, slot);
-            }
+            self.rename_tree_root(&orphaned_tree, TreeRoot::Root(slot));
             self.trees.insert(slot, orphaned_tree);
+        } else if let Some(TreeRoot::PrunedRoot(subtree_root)) =
+            self.slot_to_tree.get(&slot).copied()
+        {
+            info!(
+                "Dumping pruned slot {} of tree {} in repair",
+                slot, subtree_root
+            );
+            let mut subtree = self
+                .pruned_trees
+                .remove(&subtree_root)
+                .expect("`self.slot_to_tree` and `self.pruned_trees` must be in sync");
+
+            if subtree_root == slot {
+                // In this case we simply unprune the entire subtree
+                self.rename_tree_root(&subtree, TreeRoot::Root(subtree_root));
+                self.trees.insert(subtree_root, subtree);
+            } else {
+                let orphaned_tree = subtree.split_off(&(slot, Hash::default()));
+                self.pruned_trees.insert(subtree_root, subtree);
+                self.rename_tree_root(&orphaned_tree, TreeRoot::Root(slot));
+                self.trees.insert(slot, orphaned_tree);
+            }
```

Comment:
Is the reasoning behind this `TreeRoot::PrunedRoot` case where we unprune the tree based on the assumption that the caller is `repair_service` where we've dumped this particular `slot` and thus need to *reset* because we're not sure about the ancestry?

If so, I feel like this requires the `RepairWeight` data structure know too much about the calling logic. I think this responsibility would be better propagated up into `repair_service` where:
1. `RepairService` calls a new function `unprune()` that moves the pruned slot into the unpruned set. This can then be unit tested.
2. `RepairService` then calls `split_off`, which solely handles the orphaning logic

Reply:
The idea here is that split_off is the general purpose function to split a tree (whether it is pruned or not) and add it as an orphan to `self.trees`. This function is used whenever we're not sure about ancestory as orphan trees will have their ancestory reevaluated. This can be used by repair_service under normal dump & repair, or due to ancestor hashes repair or even for ghsa.

This case was simply an optimization, if the slot being split out is already the root of the tree no need to call `split_off` just cut and paste it to the orphan set. See the above case in `TreeRoot::Root` for `if subtree_root == slot`.

If it makes it less confusing I can combine it with the `else` arm by checking if `subtree` is empty post split and not adding it back to `self.pruned_trees`

Reply:
That makes sense! I think we can leave it as is then, and just fold your explanation of purpose of the function into the function comment as suggested here: https://github.com/solana-labs/solana/pull/29922#discussion_r1111172432